### PR TITLE
MO fallback to old path if some features are not supported yet

### PR DESCRIPTION
**Pull request by mbencer**
Ticket: 71715

### CODE REVIEW

File: model-optimizer/mo/main.py [hunk 1]
```diff
@@ -313,15 +313,17 @@ def raise_ie_not_found():
...
+def prepare_ir(argv : argparse.Namespace):
     argv = arguments_post_parsing(argv)
 
     t = tm.Telemetry()
     graph = None
     ngraph_function = None
     moc_front_end, available_moc_front_ends = get_moc_frontends(argv)
-
-    if moc_front_end:
+    apply_fallback = False
```
Comment (tomdol) [inline model-optimizer/mo/main.py:323]: Can you wrap it in something like this?

```
def use_fallback(argv):
    def extensions():
        return hasattr(argv, 'extensions') and argv.extensions is not None and len(argv.extensions) > 0
    def transformations():
        return hasattr(argv, 'transformations_config') and argv.transformations_config is not None and len(argv.transformations_config) > 0

    return extensions() or transformations()

apply_fallback = use_fallback (argv)
```

You can even call it directly in the if statement in that case.

Comment (mbencer) [reply model-optimizer/mo/main.py]: @tomdol I've used the map here in order to use a key as info for events/logs

**Review by tomdol — APPROVED** [review]

File: tools/mo/openvino/tools/mo/main.py [hunk 2]
```diff
@@ -313,21 +317,41 @@ def raise_ie_not_found():
...
 
 
-def prepare_ir(argv):
-    argv = arguments_post_parsing(argv)
+def check_fallback(argv : argparse.Namespace):
+    fallback_reasons = {}
+    fallback_reasons['extensions'] = \
+        lambda argv : hasattr(argv, 'extensions') and argv.extensions is not None and len(argv.extensions) > 0
+    fallback_reasons['transformations_config'] = \
+        lambda argv: hasattr(argv, 'transformations_config') and argv.transformations_config is not None and len(argv.transformations_config) > 0
```
Comment (popovaan) [inline tools/mo/openvino/tools/mo/main.py:325]: Please, make sure that transformations_config path or extensions paths are not sent to analytics.

**Review by popovaan — CHANGES_REQUESTED** [review]

File: tools/mo/openvino/tools/mo/main.py (hunk 2 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py:325]: The final messages contain only 'extensions' and 'transformations_config' string (the keys of `fallback_reasons` dict)

File: tools/mo/unit_tests/mo/utils/mo_fallback_test.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
+# Copyright (C) 2018-2021 Intel Corporation
```
Comment (nosovmik) [inline tools/mo/unit_tests/mo/utils/mo_fallback_test.py:1]: Not sure that these tests will work this way on CI. Probably some wrapper (like used in frontend_ngraph_test.py) would be needed

File: tools/mo/unit_tests/mo/utils/mo_fallback_test.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                f.write("[]") # json format
+            args.transformations_config = os.path.abspath(trans_config)
+        else:
+            args.transformations_config = trans_config
+
+        prepare_ir(args)
+
+        tm.Telemetry.send_event.assert_any_call('mo', 'conversion_method', expected_path)
+
+        if args.transformations_config is not None:
```
[the remark this answers is not in the document]
Comment (mbencer) [reply tools/mo/unit_tests/mo/utils/mo_fallback_test.py:156]: right, fixed

File: tools/mo/unit_tests/mo/utils/mo_fallback_test.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        args = base_args_config()
+        args.use_legacy_frontend = use_legacy
+        args.use_new_frontend = use_new_fe
+        args.input_model = "fake_elu.onnx"
+        # FakeElu is supported only by legacy path
+        Op.registered_ops['FakeElu'] = Elu
+
+        prepare_ir(args)
+
+        call_args_dict = event_call_args_to_dict(tm.Telemetry.send_event.call_args_list)
```
[the remark this answers is not in the document]
Comment (mbencer) [reply tools/mo/unit_tests/mo/utils/mo_fallback_test.py]: that's right. `event_call_args_to_dict` helper is needed in telemetry part

File: tools/mo/unit_tests/mo/utils/mo_fallback_test.py (hunk 3 above)
Comment (mbencer) [reply tools/mo/unit_tests/mo/utils/mo_fallback_test.py]: good catch, added

File: tools/mo/unit_tests/mo/frontend_ngraph_test.py [hunk 6]
```diff
@@ -59,3 +59,12 @@ def test_main_test():
 
     status = subprocess.run(args, env=os.environ)
     assert not status.returncode
+
+
+def test_mo_fallback_test():
+    setup_env()
+    args = [sys.executable, '-m', 'pytest',
+            os.path.join(os.path.dirname(__file__), 'utils/mo_fallback_test.py'), '-s']
```
Comment (nosovmik) [inline tools/mo/unit_tests/mo/frontend_ngraph_test.py:67]: should it be mo_fallback_test_actual.py? Otherwise CI may search for *_test.py pattern and will try to execute this file without proper 'setup_env'. But I see that now CI jobs already do appropriate environment setup before Model Optimizer UT, so if CI passes this can be good

**Review by nosovmik — APPROVED** [review]

File: tools/mo/unit_tests/mo/frontend_ngraph_test.py (hunk 6 above)
Comment (mbencer) [reply tools/mo/unit_tests/mo/frontend_ngraph_test.py:67]: OK, I see. You are right that tests in the current config were run two times. Changed to _acutal.py convention.

**Review by rkazants — APPROVED** [review]
Looks great to me. Thanks!

**Comment by rkazants** [comment]
@popovaan, could you please if we have an issue with messages for Telemetry?

**Comment by mbencer** [comment]
> @popovaan, could you please check if we have an issue with messages for Telemetry? I recommend to receive Anastasia's approval as well to make sure we have no user data leaks in Telemetry.

@rkazants I am processing the telemetry part separately in https://github.com/openvinotoolkit/openvino/pull/9318 just to not block this feature until we haven't decision about allowed telemetry data

**Review by popovaan — APPROVED** [review]

**Review by rkazants — CHANGES_REQUESTED** [review]
I need to review the changes first

File: tools/mo/openvino/tools/mo/main.py [hunk 7]
```diff
@@ -310,24 +319,45 @@ def raise_ie_not_found():
...
-        ngraph_function = moc_pipeline(argv, moc_front_end)
-    else:
-        t.send_event("mo", "conversion_method", "mo_legacy")
-        graph = unified_pipeline(argv)
+        try:
+            ngraph_function = moc_pipeline(argv, moc_front_end)
+            return graph, ngraph_function
+        except Exception as e:
+            fallback_reasons.append(f"frontend failure with exception: {e}")
+    if len(fallback_reasons) > 0:
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:350]: We have PDPD frontend that does not have legacy scenario so fallback for this case is not possible. Now I see that `unified_pipeline` tries to be called

File: tools/mo/openvino/tools/mo/main.py [hunk 8]
```diff
@@ -310,24 +319,45 @@ def raise_ie_not_found():
...
+        argv.use_legacy_frontend = True
+        load_extensions(argv)
+        reasons_message = ", ".join(fallback_reasons)
+        log.warning("The IR preparation was executed by the legacy MO path. "
+                    "This is a fallback scenario applicable only for some specific cases. "
+                   f"The detailed reason why fallback was executed: {reasons_message}. "
+                    "You can specify --use_legacy_frontend flag to force using the legacy MO path to avoid additional checks.")
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:357]: Also, I have question how we are going to test new frontend features (that cover the current limitations) in our validation, if it automatically fallbacks legacy frontend. For example, once we start to support extensions and transformation_config files how will we track regressions?

File: tools/mo/openvino/tools/mo/main.py [hunk 9]
```diff
@@ -310,24 +319,45 @@ def raise_ie_not_found():
...
         moc_front_end.add_extension(TelemetryExtension("mo", t.send_event, t.send_error, t.send_stack_trace))
-        ngraph_function = moc_pipeline(argv, moc_front_end)
-    else:
-        t.send_event("mo", "conversion_method", "mo_legacy")
-        graph = unified_pipeline(argv)
+        try:
+            ngraph_function = moc_pipeline(argv, moc_front_end)
+            return graph, ngraph_function
+        except Exception as e:
+            fallback_reasons.append(f"frontend failure with exception: {e}")
```
Comment (slyalin) [inline tools/mo/openvino/tools/mo/main.py:349]: It hides all errors from the new front-end. It shouldn't be. We discussed some predefined list of well described reasons for the fallback. They are listed above in the code in check_fallback. We shouldn't fallback in case of any error in the new front-end.

**Review by slyalin — CHANGES_REQUESTED** [review]

File: tools/mo/openvino/tools/mo/main.py [hunk 10]
```diff
@@ -310,24 +319,45 @@ def raise_ie_not_found():
...
+def prepare_ir(argv : argparse.Namespace):
+    argv = arguments_post_parsing(argv)
     t = tm.Telemetry()
     graph = None
     ngraph_function = None
     moc_front_end, available_moc_front_ends = get_moc_frontends(argv)
-
-    if moc_front_end:
+    fallback_reasons = check_fallback(argv)
+    if moc_front_end and len(fallback_reasons) == 0:
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:342]: in case `use_new_frontend` we should exclude fallback. The fallback is applicable only if new frontend is default and `use_new_frontend` and `use_legacy_frontend` are not specified in cmd line

File: tools/mo/openvino/tools/mo/main.py (hunk 7 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py:348]: Good point! Fixed

File: tools/mo/openvino/tools/mo/main.py (hunk 8 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py]: The only models from the current e2e validation which I've found as affected by fallback are maskrcnn and fasterrcnn (`--transformations_config` used).

They are tested now: 
- in one pipeline: without any flag (so legacy is the default path) 
- and with `--use_new_frontend` explicitly (so on new frontend without fallback - after applying https://github.com/openvinotoolkit/openvino/pull/9034#discussion_r775787984)

If the new frontend will be the default path, we can ask the validation team to run maskrcnn and fasterrcnn (without cmd flag) with expected pass result and additionally with `--use_new_frontend` (without fallback) with expected XFAIL.

@rkazants What do you think? Should I create a ticket for it?

File: tools/mo/openvino/tools/mo/main.py (hunk 9 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py:347]: @slyalin  I see your point. I've started the mail thread about it. In order to not block this PR I'm processing it here without handling the exception case.

File: tools/mo/openvino/tools/mo/main.py (hunk 10 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py]: It seems to make sense. I've changed the logic

File: tools/mo/openvino/tools/mo/utils/cli_parser.py [hunk 11]
```diff
@@ -435,10 +435,10 @@ def get_common_cli_parser(parser: argparse.ArgumentParser = None):
                               help=argparse.SUPPRESS, action=DeprecatedStoreTrue, default=False)
     common_group.add_argument("--use_new_frontend",
                               help="Use new frontend API for model processing",
-                              action='store_true', default=False)
+                              action='store_true', default=argparse.SUPPRESS)
     common_group.add_argument("--use_legacy_frontend",
                               help="Use legacy API for model processing",
-                              action='store_true', default=False)
+                              action='store_true', default=argparse.SUPPRESS)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/utils/cli_parser.py:441]: please do not suppress these options since we use them for new TF frontend (validation, etc.)

File: tools/mo/openvino/tools/mo/main.py [hunk 12]
```diff
@@ -308,24 +312,50 @@ def raise_ie_not_found():
...
+            ngraph_function = moc_pipeline(argv, moc_front_end)
+            return graph, ngraph_function
+        else: # apply fallback
+            reasons_message = ", ".join(fallback_reasons)
+            argv.use_legacy_frontend = True
+            load_extensions(argv)
+            log.warning("The IR preparation was executed by the legacy MO path. "
+                        "This is a fallback scenario applicable only for some specific cases. "
+                       f"The detailed reason why fallback was executed: not supported {reasons_message} were used. "
+                        "You can specify --use_legacy_frontend flag to force using the legacy MO path to avoid additional checks.")
```
Comment (slyalin) [inline tools/mo/openvino/tools/mo/main.py:355]: Mentioning only `--use_legacy_frontend` in case when the legacy FE has been already triggered automatically seems to be not very relevant because if `--use_legacy_frontend` is used in this case it will produce IR in the same way as without using this key except this warning message won't be printed out. Instead, mentioning `--use_new_frontend` is more useful for the customer because it forces the new behavior and shouldn't involve any logic to decide which frontend is called. I think we should provide a link to a more detailed description of what's going on with all these `use_new_frontend`/`use_legacy_frontend`.

**Review by slyalin — CHANGES_REQUESTED** [review]

File: tools/mo/openvino/tools/mo/main.py (hunk 12 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py]: done

File: tools/mo/openvino/tools/mo/utils/cli_parser.py (hunk 11 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/utils/cli_parser.py]: done

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 13]
```diff
@@ -626,4 +626,12 @@ The issue "SyntaxError: 'yield' inside list comprehension" might occur during co
 The following workarounds are suggested to resolve this issue:
 1. Use Python 3.6/3.7 to convert MXNet\* models on Windows
 2. Update MXNet: pip install mxnet=1.7.0.post2
-Note that you might have conflicts between previously installed PyPI dependencies.
\ No newline at end of file
+Note that you might have conflicts between previously installed PyPI dependencies.
+
+#### 105. What does the message "The IR preparation was executed by the legacy MO path. ..." mean? <a name="question-105"></a>
+
+For some model formats (in current state for ONNX and TensorFlow) are available two paths of IR conversion.
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:633]: please mention only ONNX, because TensorFlow is not going to be released in 2022.1.

Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:633]: the current state?

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 14]
```diff
@@ -626,4 +626,12 @@ The issue "SyntaxError: 'yield' inside list comprehension" might occur during co
...
 1. Use Python 3.6/3.7 to convert MXNet\* models on Windows
 2. Update MXNet: pip install mxnet=1.7.0.post2
-Note that you might have conflicts between previously installed PyPI dependencies.
\ No newline at end of file
+Note that you might have conflicts between previously installed PyPI dependencies.
+
+#### 105. What does the message "The IR preparation was executed by the legacy MO path. ..." mean? <a name="question-105"></a>
+
+For some model formats (in current state for ONNX and TensorFlow) are available two paths of IR conversion. 
+The old one is handled by old python implementation while the new one is realized via new c++ frontends.
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:634]: python -> Python, c++ -> C++, realized (I believe not the right word here) -> processed

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 15]
```diff
@@ -626,4 +626,12 @@ The issue "SyntaxError: 'yield' inside list comprehension" might occur during co
...
\ No newline at end of file
+Note that you might have conflicts between previously installed PyPI dependencies.
+
+#### 105. What does the message "The IR preparation was executed by the legacy MO path. ..." mean? <a name="question-105"></a>
+
+For some model formats (in current state for ONNX and TensorFlow) are available two paths of IR conversion. 
+The old one is handled by old python implementation while the new one is realized via new c++ frontends. 
+Since version 2022.1 the default path of IR conversion for ONNX models is handled via new ONNX frontend. 
+Some features like `--extensions` and `--transformations_config` are not supported yet on the new frontends. 
+If a user doesn't choose expected path of conversion explicitly (by `--use_new_frontend` or `--use_legacy_frontend` MO arguments) and some not supported pre-defined scenario is detected on the new frontend path, the IR conversion falls back to the old path.
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:637]: an user does not choose

Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:637]: some unsupported pre-defined

File: tools/mo/openvino/tools/mo/utils/cli_parser.py [hunk 16]
```diff
@@ -435,10 +435,10 @@ def get_common_cli_parser(parser: argparse.ArgumentParser = None):
                               help=argparse.SUPPRESS, action=DeprecatedStoreTrue, default=False)
     common_group.add_argument("--use_new_frontend",
                               help="Use new frontend API for model processing",
-                              action='store_true', default=False)
+                              action='store_true', default=None)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/utils/cli_parser.py:438]: please return to False

File: tools/mo/openvino/tools/mo/utils/cli_parser.py [hunk 17]
```diff
@@ -435,10 +435,10 @@ def get_common_cli_parser(parser: argparse.ArgumentParser = None):
...
     common_group.add_argument("--use_legacy_frontend",
                               help="Use legacy API for model processing",
-                              action='store_true', default=False)
+                              action='store_true', default=None)
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/utils/cli_parser.py:441]: please revert to False

File: tools/mo/openvino/tools/mo/main.py [hunk 18]
```diff
@@ -447,6 +479,12 @@ def main(cli_parser: argparse.ArgumentParser, fem: FrontEndManager, framework: s
             argv.framework = framework
         argv.feManager = fem
 
+        # Set which frontend to use by default, values should be 'new' or 'legacy'
+        argv.frontend_defaults = {
+            'onnx': 'legacy',
+            'tf': 'legacy'
+        }
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:486]: please revert this change and move to the previous place

File: tools/mo/openvino/tools/mo/main.py [hunk 19]
```diff
@@ -99,8 +99,8 @@ def get_moc_frontends(argv: argparse.Namespace):
     fem = argv.feManager
 
     # Read user flags:
-    use_legacy_frontend = argv.use_legacy_frontend
-    use_new_frontend = argv.use_new_frontend
+    use_legacy_frontend = argv.use_legacy_frontend if argv.use_legacy_frontend else False
+    use_new_frontend = argv.use_new_frontend if argv.use_new_frontend else False
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:103]: please revert it since it is equal to `False` be default

**Review by rkazants — COMMENTED** [review]
please take a look at left comments

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md (hunk 15 above)
Comment (mbencer) [reply docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:637]: I am not an expert but for me, it should be `a user` (in this case we base on how it *sounds*) but let's wait for doc team opinion ;)

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md (hunk 13 above)
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:633]: For the models in ONNX\* format, there are two available paths of IR conversion. 

(excluded TensorFlow,  referring to the comment above mentioning that  TensorFlow is not going to be released in 2022.1)

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md (hunk 14 above)
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:634]: The old one is handled by the old Python\* implementation, while the new one uses new C++ frontends.

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 20]
```diff
@@ -626,4 +626,12 @@ The issue "SyntaxError: 'yield' inside list comprehension" might occur during co
...
 2. Update MXNet: pip install mxnet=1.7.0.post2
-Note that you might have conflicts between previously installed PyPI dependencies.
\ No newline at end of file
+Note that you might have conflicts between previously installed PyPI dependencies.
+
+#### 105. What does the message "The IR preparation was executed by the legacy MO path. ..." mean? <a name="question-105"></a>
+
+For some model formats (in current state for ONNX and TensorFlow) are available two paths of IR conversion. 
+The old one is handled by old python implementation while the new one is realized via new c++ frontends. 
+Since version 2022.1 the default path of IR conversion for ONNX models is handled via new ONNX frontend.
```
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:635]: Starting from the 2022.1 version, the default IR conversion path for ONNX models is processed using the new ONNX frontend.

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 21]
```diff
@@ -626,4 +626,12 @@ The issue "SyntaxError: 'yield' inside list comprehension" might occur during co
...
-Note that you might have conflicts between previously installed PyPI dependencies.
\ No newline at end of file
+Note that you might have conflicts between previously installed PyPI dependencies.
+
+#### 105. What does the message "The IR preparation was executed by the legacy MO path. ..." mean? <a name="question-105"></a>
+
+For some model formats (in current state for ONNX and TensorFlow) are available two paths of IR conversion. 
+The old one is handled by old python implementation while the new one is realized via new c++ frontends. 
+Since version 2022.1 the default path of IR conversion for ONNX models is handled via new ONNX frontend. 
+Some features like `--extensions` and `--transformations_config` are not supported yet on the new frontends.
```
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:636]: Certain features, such as `--extensions` and `--transformations_config`, are not yet supported on the new frontends.

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md (hunk 15 above)
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:637]: The IR conversion is performed via the old path in the following cases: 
- a user does not select the new expected path of conversion explicitly by `--use_new_frontend` or `--use_legacy_frontend` MO arguments;
- an unsupported pre-defined scenario is detected on the new frontend path.

Comment (mbencer) [reply docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:637]: It rather should be:
The IR conversion is performed via the old path if a user does not select ~the new~ *any* expected path of conversion explicitly (by `--use_new_frontend` or `--use_legacy_frontend` MO arguments) and an unsupported pre-defined scenario is detected on the new frontend path.
I am not sure if the word *fallback* shouldn't be used here explicitly (it is rather common in developer slogan, but I can be wrong)

File: tools/mo/openvino/tools/mo/main.py (hunk 18 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py]: done, but I've moved it to a separate function to allow mocking in UTs

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md (hunk 15 above)
Comment (tsavina) [reply docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md]: Ok then, agreed, let's say:
The IR conversion falls back to the old path if a user does not select any expected path of conversion explicitly (by --use_new_frontend or --use_legacy_frontend MO arguments) and unsupported pre-defined scenario is detected on the new frontend path.

File: tools/mo/openvino/tools/mo/main.py [hunk 22]
```diff
@@ -308,25 +322,53 @@ def raise_ie_not_found():
...
+        fallback_reasons = check_fallback(argv)
+        if len(fallback_reasons) == 0:
+            t.send_event("mo", "conversion_method", moc_front_end.get_name() + "_frontend")
+            moc_front_end.add_extension(TelemetryExtension("mo", t.send_event, t.send_error, t.send_stack_trace))
+            moc_front_end.add_extension(ProgressReporterExtension(progress_printer(argv)))
+            ngraph_function = moc_pipeline(argv, moc_front_end)
+            return graph, ngraph_function
+        else: # apply fallback
+            reasons_message = ", ".join(fallback_reasons)
+            argv.use_legacy_frontend = True
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:361]: I think we don't need to change argv parameters because we throw warning in this case and notify an user about legacy path

File: tools/mo/openvino/tools/mo/main.py [hunk 23]
```diff
@@ -308,25 +322,53 @@ def raise_ie_not_found():
...
         import_extensions.load_dirs(argv.framework, extensions, get_front_classes)
 
-    return argv
 
+def check_fallback(argv : argparse.Namespace):
+    fallback_reasons = {}
 
-def prepare_ir(argv):
-    argv = arguments_post_parsing(argv)
+    if not any(deduce_framework_by_namespace(argv)): # no legacy path
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:329]: please add a comment:
```
Some frontend such PDPD does not have legacy path so it has no reasons to fallback
```

File: tools/mo/openvino/tools/mo/main.py [hunk 24]
```diff
@@ -308,25 +322,53 @@ def raise_ie_not_found():
...
+        return fallback_reasons
+
+    if argv.use_new_frontend or argv.use_legacy_frontend: # frontend chosen explicitly
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:332]: I think it is sufficient to have here just `if argv.use_new_frontend` because `moc_front_end` is None in case `use_legacy_frontend`.
Also please add a comment that
```
There is no possibility for fallback if an user strictly wants to use new frontend
```

**Review by rkazants — COMMENTED** [review]
Last comments. The rest looks excellent to me

File: tools/mo/openvino/tools/mo/main.py [hunk 25]
```diff
@@ -95,6 +95,15 @@ def print_argv(argv: argparse.Namespace, is_caffe: bool, is_tf: bool, is_mxnet:
     print('\n'.join(lines), flush=True)
 
 
+def get_default_frontends():
+    # Set which frontend to use by default, values should be 'new' or 'legacy'
+    default_frontends = {
+        'onnx': 'legacy',
```
Comment (rkazants) [inline tools/mo/openvino/tools/mo/main.py:101]: are we going to switch to new ONNX frontend in a separate PR?

Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py:101]: yes, it is connected with some more activities, more details in 75535

File: tools/mo/openvino/tools/mo/main.py [hunk 26]
```diff
@@ -281,13 +286,22 @@ def raise_ie_not_found():
...
+    load_extensions(argv)
+
+    return argv
+
+
+def load_extensions(argv : argparse.Namespace):
+    moc_front_end, available_moc_front_ends = get_moc_frontends(argv)
+
+    is_tf, is_caffe, is_mxnet, is_kaldi, is_onnx =\
+        deduce_framework_by_namespace(argv) if not moc_front_end else [False, False, False, False, False]
```
Comment (slyalin) [inline tools/mo/openvino/tools/mo/main.py:301]: Nit: Similar code should exists somewhere outside `load_extensions`. And it indeed exists at https://github.com/openvinotoolkit/openvino/pull/9034/files#diff-5c3563b31bdd39b460b6e4fbe9c30a3737d407038e0abea8358d6bf4550a66b5R140. Not a big deal, but why do we need to detect it twice when the first time should be enough? Can we save it somehow and reuse? The original logic in this line and implied code below is not perfect and now we have duplicated it. I see your point in isolation to a separate function but it is not a reason to duplicate the code of FW detection. With new FEs we introduce some canonical names for FW in OpenVINO space, can we detect name of FW once and the build the logic based on this detection.

File: tools/mo/openvino/tools/mo/main.py [hunk 27]
```diff
@@ -308,25 +322,53 @@ def raise_ie_not_found():
...
-    argv = arguments_post_parsing(argv)
+    if not any(deduce_framework_by_namespace(argv)): # no legacy path
+        return fallback_reasons
+
+    if argv.use_new_frontend or argv.use_legacy_frontend: # frontend chosen explicitly
+        return fallback_reasons
+
+    fallback_reasons['extensions'] = \
+        lambda argv : hasattr(argv, 'extensions') and argv.extensions is not None and len(argv.extensions) > 0 \
+            and argv.extensions != import_extensions.default_path() # extensions arg has default value
```
Comment (slyalin) [inline tools/mo/openvino/tools/mo/main.py:337]: Does it mean that I cannot activate new Extensions because it will always fallback to legacy FE?

@itikhono, please take a look. Some long time ago we discussed possibility to pass so/dll libraries with new extensions to new FE through `--extension` command line parameter. Are we going to deliver this functionality? Is it going to be a part of a separate PR?

File: tools/mo/openvino/tools/mo/main.py [hunk 28]
```diff
@@ -308,25 +322,53 @@ def raise_ie_not_found():
...
+    fallback_reasons['transformations_config'] = \
+        lambda argv: hasattr(argv, 'transformations_config') and argv.transformations_config is not None and len(argv.transformations_config) > 0
```
Comment (slyalin) [inline tools/mo/openvino/tools/mo/main.py:339]: Does it mean that I cannot pass JsonConfigExtension? @itikhono please look at that.

File: tools/mo/openvino/tools/mo/main.py (hunk 22 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py:361]: it is required by `load_extensions` method

File: tools/mo/openvino/tools/mo/main.py (hunk 26 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py]: @slyalin fixed in https://github.com/openvinotoolkit/openvino/pull/9034/commits/2d0c78cd7ebe3c94207fd2eb349c1daeb3f28757

File: tools/mo/openvino/tools/mo/main.py (hunk 22 above)
Comment (mbencer) [reply tools/mo/openvino/tools/mo/main.py]: finally removed, `load_extensions` parameterized in other way

File: tools/mo/openvino/tools/mo/main.py (hunk 28 above)
Comment (itikhono) [reply tools/mo/openvino/tools/mo/main.py:336]: @slyalin I guess we don't have any implemented transformations in c++ part to run via JsonConfig, doesn't that mean we should always use a fallback? 
And I think we already don't have time to implement such transformations in 2022.1R

File: tools/mo/openvino/tools/mo/main.py (hunk 27 above)
Comment (itikhono) [reply tools/mo/openvino/tools/mo/main.py:334]: >> Does it mean that I cannot activate new Extensions because it will always fallback to legacy FE?
Looks like that and I think it's ok for this PR

Yes, `--extension`  with so/dll looks useful, when we want to combine extensions, e.g. ConversionExtensions, into a separate module and loading them via MO interface. 
I will create a new ticket in extension epic, but if it's already exists please let me know.

File: tools/mo/openvino/tools/mo/main.py [hunk 29]
```diff
@@ -95,6 +95,15 @@ def print_argv(argv: argparse.Namespace, is_caffe: bool, is_tf: bool, is_mxnet:
     print('\n'.join(lines), flush=True)
 
 
+def get_default_frontends():
+    # Set which frontend to use by default, values should be 'new' or 'legacy'
+    default_frontends = {
+        'onnx': 'legacy',
+        'tf': 'legacy'
```
[the remark this answers is not in the document]
Comment (rkazants) [reply tools/mo/openvino/tools/mo/main.py:102]: we will have separate PR for sure. There is a P1 ticket for this

**Review by rkazants — APPROVED** [review]

File: tools/mo/openvino/tools/mo/main.py (hunk 27 above)
Comment (slyalin) [reply tools/mo/openvino/tools/mo/main.py:334]: @itikhono, agree. But we should do this soon, otherwise we won't have possibility to pass new extensions to FEs through MO.

File: tools/mo/openvino/tools/mo/main.py (hunk 28 above)
Comment (slyalin) [reply tools/mo/openvino/tools/mo/main.py:336]: Discussed with @itikhono. JsonConfigExtension is a legal replacement of old config-enabled way in MO but for new FEs. It seems to be tested enough to enable it here. Regardless of fact that we don't have our own implemented JsonConfigExtension's. We don't really want to have them inside the product and leaving this way for advanced users as an extreme measure. @mbencer, please put more fine-grain condition here to let new transformation config pass to new FEs. New config can be detected by identifying a field `library` in json file at the level of each transformation. Please consult with @itikhono for details. I would prefer to have this in a separate PR to avoid blocking of validation the fallback itself. But we must have that. Otherwise all stuff made for JsonConfigExtension cannot be activated by a user.

File: tools/mo/openvino/tools/mo/main.py (hunk 27 above)
Comment (slyalin) [reply tools/mo/openvino/tools/mo/main.py:334]: AR for @itikhono to continue on that.

**Review by itikhono — APPROVED** [review]

**Review by slyalin — APPROVED** [review]
@mbencer, @itikhono, please follow up on two raised topics.
